drivers/ethernet/phy/phy_mii.c

```diff
@@ -294,9 +308,45 @@ static void monitor_work_handler(struct k_work *work)
 		invoke_link_cb(dev);
 	}
 
-	/* Submit delayed work */
-	k_work_reschedule(&data->monitor_work,
-			  K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+	if (rc == -EINPROGRESS) {
+		/* Check for autonegotiation completion */
+		k_work_reschedule(&data->autoneg_work,
+				  K_MSEC(MII_AUTONEG_POLL_INTERVAL_MS));
+	} else {
+		/* Submit delayed work */
+		k_work_reschedule(&data->monitor_work,
+				  K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+	}
+}
+
+static void autoneg_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct phy_mii_dev_data *const data =
+		CONTAINER_OF(dwork, struct phy_mii_dev_data, autoneg_work);
+	const struct device *dev = data->dev;
+	int rc;
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	rc = check_autonegotiation_completion(dev);
+
+	k_sem_give(&data->sem);
```

Comment:
There's a discussion ongoing with regards to making blocking calls in the system work queue: https://github.com/zephyrproject-rtos/zephyr/pull/87522

This could be such a case, where I'm not sure we'd want to wait? Would it be an option to take the semaphore with `K_NO_WAIT`, and reschedule the work if the semaphore isn't available?

Reply:
Lots of interesting context in that discussion, thanks.

Reply:
I've made that `k_sem_take()` and the one in `monitor_work_handler()` non-blocking.